paper_framework_sources/p1898_forward_progress_delegation.md

```diff
@@ -1,28 +1,35 @@
+This design enables delegation of forward progress in asynchronous algorithms and enables well-defined propagation of execution resources through an asynchronous work chain.
+This work is based in part on a discussion of the `DeferredExecutor` that works with [folly's](https://github.com/facebook/folly/) [SemiFuture](https://github.com/facebook/folly/blob/master/folly/futures/Future.h) and how folly's coroutines propagate executors as well to ensure that work runs on predictable underlying execution resources.
 
-In this paper `executor` refers to the general concept as expressed in the title of [@P0443R11] and the long standing executors effort.
-The concrete concepts used may be `executor`, `scheduler` or `sender` depending on the context.
+In this paper `executor` refers to the general concept as expressed in the title of [@P0443R13] and the long standing executors effort.
+The concrete concepts used may satisfy `executor`, `scheduler` or `sender` depending on the context.
 
-Building on the definitions in [@P0443R11] and [@P1660R0], we start with the definition of a `receiver` we propose adding a concept `scheduler_provider` that allows a `sender` to require that a `receiver` passed to it is able to provide an executor, or more accurately a `scheduler` in the language of [@P0443R11].
-This concept offers a `get_scheduler` customization point that allows a `scheduler` to be propagated from downstream to upstream asynchronous tasks.
-A given `scheduler`, and therefore any `sender` created off that scheduler may require the availability of such a customization on `recivers`s passed to the `submit` operation.
+Building on the definitions in [@P0443R13] and [@P1897R3], we start with the definition of a `receiver` we propose adding a concept `scheduler_provider` that allows a `sender` to require that a `receiver` passed to it is able to provide an executor, or more accurately a `scheduler` in the language of [@P0443R13].
+This concept offers a `get_scheduler` customization point that allows a `scheduler` to be propagated from downstream to upstream asynchronous tasks, or alternatively from upstream to downstream.
+A given `scheduler`, and therefore any `sender` created off that scheduler may require the availability of such a customization on `receivers`s passed to the `submit` operation.
 
 A given `sender` may delegate any tasks it is unable to perform to the scheduler it acquires from the downstream `Callback`.
+A given `sender` alternatively might always reschedule work to an underlying `scheduler`, and in situations where the chosen `scheduler` is ambiguous, for example where the output work runs from a `when_all` operation, that `sender` might require that downstream work provide a `scheduler` for it to reschedule onto to remove nondeterminism in where work runs.
```

Comment:
Yes, I think that's right. The customisation of when_all does a transition in your example. From the POV of the user, though, that means that when_all might do a transition. 

What I'm saying here though is that if I call:
`when_all(transform(on(s1, sch1), f1), transform(on(s2, sch2), f2))`

sch1 and sch2 do not match, so the scheduler that completes when_all is not deterministic. In that case we might require when_all to return a semi-sender, in folly parlance, such that it requires a scheduler_provider.

So:
`connect(when_all(transform(on(s1, sch1), f1), transform(on(s2, sch2), f2)), some-non-provider-receiver);
`
would fail to compile but:

`on(when_all(transform(on(s1, sch1), f1), transform(on(s2, sch2), f2)), sch3)`


would be fine, because on is going to call connect with a scheduler_provider

Reply:
If you want to have when_all() complete on a particular execution context, you can just apply via() to the result of when_all()?

`via(when_all(transform(on(s1, sch1), f1), transform(on(s2, sch2), f2)), sch3)`

Note, however, that in this example it's not guaranteed that f1/f2 will be executed on sch1/sch2 respectively. f1/f2 will be executed on whatever context s1/s2 complete on.

As I understand the on() algorithm as currently specified, it only controls the execution context that an operation starts on and injects the scheduler as the get_scheduler() context for child receivers. It doesn't seem to directly control anything about the context that the operation completes on.

Reply:
Yes, you can "just" apply via. The point here is that when_all might be defined (and some algorithms will be defined) to *require* that that "via" be there, so that there is some well-defined place to run work. As a SemiFuture does.